File: legacy/python/spider_claw/core/http.py
"""HTTP 客户端工厂：统一处理代理配置，并规避 httpx 对环境变量的解析缺陷。

背景（真实踩坑）：
    httpx 在 ``trust_env=True`` 时会解析 ``*_PROXY`` / ``NO_PROXY`` 环境变量，
    并在**构造 Client 时**就把它们转成 mount 规则。常见的 ``NO_PROXY`` 写法
    ``...::1/128...``（IPv6 CIDR）会生成 ``all://[::1/128]`` 这样的 pattern，
    解析时直接抛 ``httpx.InvalidURL("Invalid port: ':1'")`` —— 此时连
    ``httpx.AsyncClient(proxy=None)`` 都无法构造（``proxy=None`` 只表示"不指定代理"，
    仍会读取环境变量），整个抓取流程在建客户端这一步就崩掉。

因此这里统一提供两个工厂函数：
    * 直连（``proxy=None``）→ 真正关闭 env 代理（``trust_env=False``）；
    * 走环境代理 → 先净化 ``NO_PROXY``，仍解析失败则告警并降级为直连，绝不中断抓取。
"""
from __future__ import annotations

import os
import re
from contextlib import contextmanager

import httpx

# 哨兵：表示"使用 httpx 默认行为（读取环境变量中的代理）"
_USE_ENV_PROXY = object()

_NO_PROXY_KEYS = ("NO_PROXY", "no_proxy")
# IPv6 CIDR，如 ::1/128 —— httpx 会生成 all://[::1/128] 并解析失败
_IPV6_CIDR_RE = re.compile(r"^([0-9A-Fa-f:]*:[0-9A-Fa-f:]+)/\d+$")


def clean_no_proxy(value: str) -> str:
    """净化 NO_PROXY：`::1/128` → `::1`（httpx 无法解析带掩码的 IPv6）。

    其它条目（IPv4 CIDR、域名、`*` 等）原样保留。
    """
    parts: list[str] = []
    for raw in (value or "").split(","):
        host = raw.strip()
        if not host:
            continue
        m = _IPV6_CIDR_RE.match(host)
        if m:
            host = m.group(1)
        parts.append(host)
    return ",".join(parts)


@contextmanager
def _safe_proxy_env():
    """临时净化 NO_PROXY/np_proxy，退出时还原。"""
    saved = {k: os.environ.get(k) for k in _NO_PROXY_KEYS}
    for key, value in saved.items():
        if value:
            cleaned = clean_no_proxy(value)
            if cleaned != value:
                os.environ[key] = cleaned
    try:
        yield
    finally:
        for key, value in saved.items():
            if value is None:
                os.environ.pop(key, None)
            else:
                os.environ[key] = value


def _client_kwargs(proxy) -> dict:
    if proxy is None:
        # 直连：连带关闭环境变量代理，避免被畸形 NO_PROXY 弄崩
        return {"proxy": None, "trust_env": False}
    if proxy is _USE_ENV_PROXY:
        return {"trust_env": True}
    return {"proxy": proxy, "trust_env": False}


def _build(factory, proxy, kwargs):
    """构造 client；环境代理不可解析时告警并降级为直连。"""
    common = {**_client_kwargs(proxy), **kwargs}
    if proxy is None:
        return factory(**common)

    with _safe_proxy_env():
        try:
            return factory(**common)
        except httpx.InvalidURL as exc:  # 畸形代理配置（如 ::1/128）
            message = str(exc)

    print(f"  ⚠ 代理配置无法解析（{message}），本次回退为直连")
    return factory(proxy=None, trust_env=False, **kwargs)


def async_client(proxy=_USE_ENV_PROXY, **kwargs) -> httpx.AsyncClient:
    """构造 AsyncClient。proxy=None 表示强制直连（忽略 *_PROXY 环境变量）。"""
    return _build(httpx.AsyncClient, proxy, kwargs)


def sync_client(proxy=_USE_ENV_PROXY, **kwargs) -> httpx.Client:
    """构造同步 Client。语义同 :func:`async_client`。"""
    return _build(httpx.Client, proxy, kwargs)

File: src/core/http.ts
/**
 * HTTP 客户端：统一代理配置、字符集解码与超时。
 *
 * 代理语义（与 Python 版保持一致，但实现更简）：
 * - `direct`（默认）：**不读任何代理环境变量**。Node 的 fetch/undici 默认就不解析
 *   `HTTP_PROXY` 等变量，因此「直连」天然干净，不存在 Python httpx 那种
 *   "构造 Client 时就因畸形 NO_PROXY（如 ::1/128）抛 InvalidURL" 的问题；
 * - `env`：使用 undici 的 `EnvHttpProxyAgent` 读取 `HTTP(S)_PROXY` / `NO_PROXY`；
 *   构造前先净化 `NO_PROXY`（IPv6 CIDR → 去掉掩码），仍失败则告警并降级为直连；
 * - 其它字符串：视为显式代理 URL。
 */
import {
  EnvHttpProxyAgent,
  ProxyAgent,
  fetch as undiciFetch,
  type Dispatcher,
} from 'undici';

import { logger } from './logger.js';

/** 代理配置：直连 / 读取环境变量 / 显式代理 URL。 */
export type ProxyConfig = 'direct' | 'env' | (string & {});

const NO_PROXY_KEYS = ['NO_PROXY', 'no_proxy'] as const;
/** IPv6 CIDR（如 ::1/128）——部分 HTTP 客户端无法解析带掩码的 IPv6。 */
const IPV6_CIDR_RE = /^([0-9A-Fa-f:]*:[0-9A-Fa-f:]+)\/\d+$/;

/**
 * 净化 NO_PROXY：`::1/128` → `::1`。
 *
 * 其它条目（IPv4 CIDR、域名、通配符 `*`、`*.local` 等）原样保留。
 */
export function cleanNoProxy(value: string): string {
  const parts: string[] = [];
  for (const raw of (value ?? '').split(',')) {
    const host = raw.trim();
    if (!host) continue;
    const matched = IPV6_CIDR_RE.exec(host);
    parts.push(matched?.[1] ?? host);
  }
  return parts.join(',');
}

/** 在「净化后的 NO_PROXY」环境下同步执行构造逻辑，退出时还原环境变量。 */
function withCleanedNoProxy<T>(fn: () => T): T {
  const saved = NO_PROXY_KEYS.map((key) => [key, process.env[key]] as const);
  for (const [key, value] of saved) {
    if (value) {
      const cleaned = cleanNoProxy(value);
      if (cleaned !== value) process.env[key] = cleaned;
    }
  }
  try {
    return fn();
  } finally {
    for (const [key, value] of saved) {
      if (value === undefined) delete process.env[key];
      else process.env[key] = value;
    }
  }
}

export interface DispatcherResolution {
  /** `undefined` 表示使用 undici 默认 dispatcher（即直连）。 */
  dispatcher: Dispatcher | undefined;
  /** 是否因为代理配置不可用而回退到直连。 */
  degraded: boolean;
  reason?: string;
}

/** 根据代理配置构造 dispatcher；任何解析失败都降级为直连而不是抛错。 */
export function resolveDispatcher(proxy: ProxyConfig = 'direct'): DispatcherResolution {
  if (proxy === 'direct') return { dispatcher: undefined, degraded: false };
  try {
    if (proxy === 'env') {
      return { dispatcher: withCleanedNoProxy(() => new EnvHttpProxyAgent()), degraded: false };
    }
    return { dispatcher: new ProxyAgent(proxy), degraded: false };
  } catch (e) {
    return { dispatcher: undefined, degraded: true, reason: (e as Error).message };
  }
}

export interface RequestOptions {
  headers?: Record<string, string>;
  /** 便捷设置 Referer（部分站点图片 / 文章有防盗链）。 */
  referer?: string;
  timeoutMs?: number;
}

export interface RawResponse {
  bytes: Uint8Array;
  contentType: string | null;
  status: number;
}

/**
 * 按 `Content-Type` → `<meta charset>` → UTF-8 的顺序解码响应体。
 *
 * 说明：undici 的 `Response.text()` 恒定按 UTF-8 解码，会读乱 GBK 等旧编码的
 * 静态博客；这里显式嗅探字符集，行为对齐 Python 版 httpx 的 `r.text`。
 */
export function decodeBody(bytes: Uint8Array, contentType: string | null): string {
  const fromHeader = /charset=["']?([\w-]+)/i.exec(contentType ?? '')?.[1];
  const sniff = new TextDecoder('utf-8').decode(bytes.subarray(0, 2048));
  const fromMeta = /<meta[^>]+charset=["']?([\w-]+)/i.exec(sniff)?.[1];
  const charset = (fromHeader ?? fromMeta ?? 'utf-8').toLowerCase();
  if (charset === 'utf-8' || charset === 'utf8') {
    return new TextDecoder('utf-8').decode(bytes);
  }
  try {
    return new TextDecoder(charset).decode(bytes);
  } catch {
    // 未知字符集：退回 UTF-8，绝不因解码细节中断抓取
    return new TextDecoder('utf-8').decode(bytes);
  }
}

/** 轻量 HTTP 客户端（基于 undici fetch，自带 redirect / 超时 / 代理）。 */
export class HttpClient {
  readonly proxy: ProxyConfig;
  private readonly dispatcher: Dispatcher | undefined;

  constructor(proxy: ProxyConfig = 'direct') {
    this.proxy = proxy;
    const resolved = resolveDispatcher(proxy);
    this.dispatcher = resolved.dispatcher;
    if (resolved.degraded) {
      logger.warn(`  ⚠ 代理配置无法解析（${resolved.reason}），本次回退为直连`);
    }
  }

  private buildInit(options: RequestOptions): Parameters<typeof undiciFetch>[1] {
    const headers: Record<string, string> = { ...(options.headers ?? {}) };
    if (options.referer) headers['Referer'] = options.referer;
    return {
      redirect: 'follow',
      headers,
      signal: AbortSignal.timeout(options.timeoutMs ?? 20_000),
      ...(this.dispatcher ? { dispatcher: this.dispatcher } : {}),
    };
  }

  async get(url: string, options: RequestOptions = {}): Promise<RawResponse> {
    const res = await undiciFetch(url, this.buildInit(options));
    if (!res.ok) {
      throw new Error(`HTTP ${res.status} ${res.statusText} (${url})`);
    }
    return {
      bytes: new Uint8Array(await res.arrayBuffer()),
      contentType: res.headers.get('content-type'),
      status: res.status,
    };
  }

  /** 抓取文本（自动按响应头 / meta 嗅探字符集）。 */
  async getText(url: string, options: RequestOptions = {}): Promise<string> {
    const { bytes, contentType } = await this.get(url, options);
    return decodeBody(bytes, contentType);
  }
}
